File: C++/utils/ByteReader.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <memory>
#include <cstring>
#include <stdexcept>
#include <functional>

class ByteReader
{
private:
    const std::vector<uint8_t> &buffer;
    size_t position = 0;

public:
    ByteReader(const std::vector<uint8_t> &data) : buffer(data) {}

    std::vector<uint8_t> getBuffer() const
    {
        return std::vector<uint8_t>(buffer.begin() + position, buffer.end());
    }

    uint8_t readByte()
    {
        if (position >= buffer.size())
        {
            throw std::runtime_error("Buffer overflow");
        }
        return buffer[position++];
    }

    // Read in big-endian format to match Java
    int32_t readInt()
    {
        int32_t value = 0;
        value |= (static_cast<int32_t>(readByte()) << 24);
        value |= (static_cast<int32_t>(readByte()) << 16);
        value |= (static_cast<int32_t>(readByte()) << 8);
        value |= static_cast<int32_t>(readByte());
        return value;
    }

    int64_t readLong()
    {
        int64_t value = 0;
        value |= (static_cast<int64_t>(readByte()) << 56);
        value |= (static_cast<int64_t>(readByte()) << 48);
        value |= (static_cast<int64_t>(readByte()) << 40);
        value |= (static_cast<int64_t>(readByte()) << 32);
        value |= (static_cast<int64_t>(readByte()) << 24);
        value |= (static_cast<int64_t>(readByte()) << 16);
        value |= (static_cast<int64_t>(readByte()) << 8);
        value |= static_cast<int64_t>(readByte());
        return value;
    }

    double readDouble()
    {
        int64_t bits = readLong();
        double value;
        memcpy(&value, &bits, sizeof(double));
        return value;
    }

    std::string readString()
    {
        int32_t length = readInt();
        std::string str;
        str.reserve(length);
        for (int i = 0; i < length; i++)
        {
            str += static_cast<char>(readByte());
        }
        return str;
    }
};

File: Server/utils/ByteReader.java
package Server.utils;

/**
 * ByteReader class provides methods to read primitive types and strings from a
 * byte array.
 * It is used during deserialization to extract data from a serialized byte
 * array.
 */
public class ByteReader {
    private byte[] buffer; // The byte array to read from
    private int position; // Current position in the byte array

    /**
     * Constructor to initialize the ByteReader with a byte array.
     * 
     * @param data The byte array to read from.
     */
    public ByteReader(byte[] data) {
        this.buffer = data;
        this.position = 0;
    }

    /**
     * Read a single byte from the buffer.
     * 
     * @return The byte read.
     */
    public byte readByte() {
        return buffer[position++];
    }

    /**
     * Read an integer (4 bytes) from the buffer.
     * 
     * @return The integer read.
     */
    public int readInt() {
        /*
         * This function retrieves Integers from the buffer.
         * 
         * Left shift operations are necessary because:
         * The data is stored in Big Endian format. (defined in ByteBuffer.java)
         * Left shift will bring the MSB back to the correct position
         * An int in Java is 4 bytes (32 bits).
         * Since a byte can only hold 8 bits, we need to extract each byte from the
         * buffer
         */
        return ((readByte() & 0xFF) << 24) |
                ((readByte() & 0xFF) << 16) |
                ((readByte() & 0xFF) << 8) |
                (readByte() & 0xFF);
    }

    /**
     * Read a long (8 bytes) from the buffer.
     * 
     * @return The long read.
     */
    public long readLong() {
        return ((long) (readByte() & 0xFF) << 56) |
                ((long) (readByte() & 0xFF) << 48) |
                ((long) (readByte() & 0xFF) << 40) |
                ((long) (readByte() & 0xFF) << 32) |
                ((long) (readByte() & 0xFF) << 24) |
                ((long) (readByte() & 0xFF) << 16) |
                ((long) (readByte() & 0xFF) << 8) |
                (readByte() & 0xFF);
    }

    /**
     * Read a double (8 bytes) from the buffer.
     * 
     * @return The double read.
     */
    public double readDouble() {
        long bits = readLong();
        return Double.longBitsToDouble(bits);
    }

    /**
     * Read a string from the buffer.
     * 
     * @return The string read.
     */
    public String readString() {
        int length = readInt();
        byte[] bytes = new byte[length];
        for (int i = 0; i < length; i++) {
            bytes[i] = readByte();
        }
        return new String(bytes);
    }
}
